Add information about CI/CD servers (#1243)
epiphany-platform/epiphany: docs/design-docs/cicd-server/comparision_cicd.md

__jenkins-x__: possible looks ok (Tekton)

__tekton__: pipelines are CRD so can be only as a code

### Build in pods

__gocd__: Elastic agent concepts. Can create many groups (probably on different clusters - not tested yet) and assigned them to proper pipelines

__jenkins__: plugin for building in kubernetes

__jenkins-x__: building in pods in cluster jenkins-x is installed. Possible to install many jenkins-x servers (according to documentation per each team in different namespace). Able to run in multi cluster mode

__tekton__: building in cluster easily. Not possible to build on different server - but I didn't any sence in that use case. Possible to deploy on other kubernetes service.

### Secrets
__gocd__: Plugins for secrets from: hashicorp vault, kubernetes secrets, file based

__jenkins__: plugins for many options: hashicorp vault, kubernetes secrets, internal secrets, etc

__jenkins-x__: Providers for secrets from: hashicorp vault, kubernetes secrets

__tekton__: Use secrets from kubernetes so everything what is inside kubernetes can be read

### Environment varaibles:

__gocd__: multiple level of variables: environment, pipeline, stage, job

> __jenkins__: environment variables can be override

Reviewer: ```suggestion
__jenkins__: environment variables can be overriden
```

__jenkins-x__: Didn't find any information but expect it will not be worst than in gocd

__tekton__: You can read env variables from any config map so this is kind of overriding.

### Plugins

__gocd__: not big number of plugins (but is this really bad?) but very of them really usefull (LDAP, running in pods, vault, k8s secrets, docker registry, push to S3, slack notification, etc)

> __jenkins__: many, many maaaaaaaaaaaany plugins. But if there is too much of them they start making seeeeeeeeeerious issues. Each plugin has different quality and each can breake the server and has its own security issues so we have to be very careful with them.

Reviewer: ```suggestion
__jenkins__: many plugins. But if there is too much of them they start making serious issues. Each plugin has different quality and each can breake the server and has its own security issues so we have to be very careful with them.
```

__jenkins-x__: plugins are called app. There are few of them and this app are helm charts.
Jenkins-x uses embeded nexus, chartmuseum and monocular services. I don't know if the is option to get rid of them.

__tekton__: tekton itself is kind of plugin for building. You can create whatever you want in different pod and get it.

## Personal conclusion
__gocd__:
* This looks like really good CI/CD central server which can be use by many teams.

> * Really mature application. Older release on github from Nov 2014. Accordind to wiki first release in 2007!

Reviewer: ```suggestion
* Really mature application. Older release on github from Nov 2014. According to wiki first release in 2007.
```

* very intuitive
* Working really good in kubernetes
* Good granuality of permission.
* Good documentation
* Small amount of help in Internet (compare to jenkins)
* Small community

> GoCD can be easily set up for our organizations. Adding new customers should be big deal. Working with is very intuitive - old school concept of CICD.

Reviewer: Adding customers should or shouldn't be big deal?

Author: fix thanks for that.

__jenkins__:
* Production ready
* The most search CI/CD tool in google - so almost every case is describe somwhere
* Very simple
* Working very good in kubernetes
* After using it for some time pipelines are getting bigger and harder to maintain

> * Goo granuality of permission

Reviewer: ```suggestion
* Good granuality of permission
```

* XML configuration for many plugins
* Big amount of information in Internet
* Big community

> The most popular CI/CD tool. Small and simple. You can do everything as a code or by GUI - which is not good because it's temptenion to fix it right now and then probably do not put it to repository. A lot of plugins which and each of them is single point of failure. Hard to configure some plugin as a code - but still possible.

Reviewer: ```suggestion
The most popular CI/CD tool. Small and simple. You can do everything as a code or by GUI - which is not good because it's temptation to fix it right now and then probably do not put it to repository. A lot of plugins which and each of them is single point of failure. Hard to configure some plugin as a code - but still possible.
```

This project is under CDFoundation and has a big community which is really good.
My personal choice.

## Another concept CI and CD tool

Use separate tools for Continious Integration and Continious Deployment. In this concept I recognized [Tekton](tekton.dev) for building and [ArgoCD](https://argoproj.github.io/argo-cd/) for delivery purposes.

### ArgoCD

In ArgoCD you can easily deploy one of your applications described as kubernetes resources into one of your kubernetes clusters.
In that case recommended option is to have two repos one for code and one for configuration. Thanks to that you can easily separate code from configuration. It also works with one repo where you keep code and configuration in one repo.

When Argo detect changes in configuration it runs new configuration on cluster. It's simple like that.

#### User management

Possible to use: local users, SSO with Bundled Dex OIDC provider, SSO with Existing OIDC provider

#### Secrets
* Bitnami Sealed Secrets
* Godaddy Kubernetes External Secrets
* Hashicorp Vault
* Banzai Cloud Bank-Vaults
* Helm Secrets
* Kustomize secret generator plugins
* aws-secret-operator
* KSOPS

#### Conclusion

> ArgoCCD looks very good if you have a really big number of clusters you are managing. Thanks to that you can deploy whatever you want wherever you need. But this is needed for really for big scale.

Reviewer: ```suggestion
ArgoCD looks very good if you have a really big number of clusters you are managing. Thanks to that you can deploy whatever you want wherever you need. But this is needed for really for big scale.
```
